### PR TITLE
vcsim: use linked list for EventHistoryCollector

### PR DESCRIPTION
**Pull request by dougm**
PR #2134 changed EventManager's history ring buffer to a linked list.
Doing the same for EventHistoryCollector as it simplifies the code and
avoid having to lock the EventManager in the Read{Next,Prev}Events methods.

### CODE REVIEW

File: simulator/event_manager.go [hunk 1]
```diff
@@ -370,25 +364,24 @@ func (c *EventHistoryCollector) SetCollectorPageSize(ctx *Context, req *types.Se
...
+	c.page = list.New()
+	ctx.WithLock(c.m, c.fillPage)
 
 	body.Res = new(types.SetCollectorPageSizeResponse)
 	return body
 }
 
 func (c *EventHistoryCollector) ResetCollector(ctx *Context, req *types.ResetCollector) soap.HasFault {
-	c.pos = c.m.history.Back()
+	c.pos = c.page.Back()
```
Comment (embano1) [inline simulator/event_manager.go:376]: Having a hard time understanding the intended behavior of this method reading the docs:

```
Moves the "scrollable view" to the item immediately preceding the "viewable latest page". 
If you use "readPrev", ReadPreviousTasks or ReadPreviousEvents, all items are retrieved from the 
newest item to the oldest item.
```

Is this line correct based on the above method description?

File: simulator/event_manager.go [hunk 2]
```diff
@@ -370,25 +364,24 @@ func (c *EventHistoryCollector) SetCollectorPageSize(ctx *Context, req *types.Se
...
 
 	return &methods.ResetCollectorBody{
 		Res: new(types.ResetCollectorResponse),
 	}
 }
 
 func (c *EventHistoryCollector) RewindCollector(ctx *Context, req *types.RewindCollector) soap.HasFault {
```
Comment (embano1) [inline simulator/event_manager.go:383]: Assuming these methods are called via lock (as discussed earlier), so they're thread-safe right?

**Review by embano1 — COMMENTED** [review]
Just two questions for clarification.

Verified the behavior with VEBA (VMware Event Router) and a binary using EventManager against this updated vcsim and events are retrieved as expected.

Just noticed that we don't use any filterspec when creating a collector, filed an issue for this: https://github.com/vmware/govmomi/issues/2240

File: simulator/event_manager.go (hunk 1 above)
Comment (dougm) [reply simulator/event_manager.go:376]: Yes. When you call `ResetCollector()` pos is set to the back of the list, and new events are added to the back of the list.  After the Reset call, Read**Next**Events() will return `nil` until new events come in.
But you'll be able to read the entire history with `ReadPreviousEvents()` after calling Reset.

`RewindCollector()` is the opposite: Read**Prev**Events() will return `nil`, but can read the entire history with`ReadNextEvents()`

File: simulator/event_manager.go (hunk 2 above)
Comment (dougm) [reply simulator/event_manager.go:383]: Yes 👍

File: simulator/event_manager.go (hunk 1 above)
Comment (embano1) [reply simulator/event_manager.go:376]: Thx, I found the name `Reset` a bit confusing :)

**Comment by embano1** [comment]
LGTM

**Review by embano1 — APPROVED** [review]

**Pull request merged**